tests/generative/variable_strategies.py

```diff
@@ -171,24 +186,72 @@ def ge(draw, _type, frame):
+    def subtract(type_, frame):
+        return binary_operation(type_, frame, Function.Subtract)
 
-    @st.composite
-    def multiply(draw, type_, frame):
-        return draw(binary_operation(type_, frame, Function.Multiply))
+    def multiply(type_, frame):
+        return binary_operation(type_, frame, Function.Multiply)
+
+    def date_add_years(type_, frame):
+        return binary_operation_with_types(type_, int, frame, Function.DateAddYears)
+
+    def date_add_months(type_, frame):
+        return binary_operation_with_types(type_, int, frame, Function.DateAddMonths)
+
+    def date_add_days(type_, frame):
+        return binary_operation_with_types(type_, int, frame, Function.DateAddDays)
+
+    def date_difference_in_years(type_, frame):
+        return binary_operation(datetime.date, frame, Function.DateDifferenceInYears)
+
+    def date_difference_in_months(type_, frame):
+        return binary_operation(datetime.date, frame, Function.DateDifferenceInMonths)
+
+    def date_difference_in_days(type_, frame):
+        return binary_operation(datetime.date, frame, Function.DateDifferenceInDays)
 
     @st.composite
     def binary_operation(draw, type_, frame, operator_func):
         # A strategy for operations that take lhs and rhs arguments of the
         # same type
-        lhs = draw(series(type_, draw(one_row_per_patient_frame_or(frame))))
-        rhs = draw(series(type_, draw(one_row_per_patient_frame_or(frame))))
+        return draw(binary_operation_with_types(type_, type_, frame, operator_func))
+
+    @st.composite
+    def binary_operation_with_types(draw, lhs_type, rhs_type, frame, operator_func):
+        # A strategy for operations that take lhs and rhs arguments with specified lhs
+        # and rhs types (which may be different)
+
+        # A binary operation has 2 inputs, which are
+        # 1) A series drawn from the specified frame
+        # 2) one of:
+        #    a) A series drawn from the specified frame
+        #    b) A series drawn from any one-row-per-patient-frame
+        #    c) A series that is a Value
+
+        # first pick an "other" input series (i.e. #2 above), either a value series
+        # or a series drawn from a frame
+        other_series = draw(st.sampled_from([value, series]))
+        # Now pick a frame for the series to be drawn from
+        # The other frame will either be a new one-row-per-patient-frame or this frame
+        # (Note if the other_series is a value, the frame will be ignored)
+        other_frame = draw(st.one_of(one_row_per_patient_frame(), st.just(frame)))
+
+        # Pick the order of the lhs and rhs inputs built from the two frames and
+        # associated strategies
+        lhs_frame, lhs_input, rhs_frame, rhs_input = draw(
+            st.sampled_from(
+                [
+                    (frame, series, other_frame, other_series),
+                    (other_frame, other_series, frame, series),
+                ]
+            )
+        )
```

Comment:
Or see Dave's alternative suggestion.

Reply:
Unless it's more efficient, I prefer this one for clarity

Reply:
:+1: